stdlib/format.mli

```diff
@@ -44,83 +44,163 @@
+ - when the material for a pretty-printing box has been printed, call
+   [close_box ()] to close the box;
+ - at the end of pretty-printing, flush the pretty-printer to display all
+   the remaining material, e.g. evaluate [print_newline ()].
+
+   The behavior of pretty-printing commands is unspecified
+   if there is no open pretty-printing box. Each box open via
    one of the [open_] functions below must be closed using [close_box]
    for proper formatting. Otherwise, some of the material printed in the
    boxes may not be output, or may be formatted incorrectly.
 
-   In case of interactive use, the system closes all opened boxes and
-   flushes all pending text (as with the [print_newline] function)
-   after each phrase. Each phrase is therefore executed in the initial
-   state of the pretty-printer.
+   In case of interactive use, each phrase is executed in the initial state
+   of the standard pretty-printer: after each phrase execution, the
+   interactive system closes all open pretty-printing boxes, flushes all
+   pending text, and resets the standard pretty-printer.
+
+   Warning: mixing calls to pretty-printing functions of this module with
+   calls to {!Pervasives} low level output functions is error prone.
+
+   The pretty-printing functions output material that is delayed in the
+   pretty-printer queue and stacks in order to compute proper line
+   splitting. In contrast, basic I/O output functions write directely in
+   their output device. As a consequence, the output of a basic I/O function
+   may appear before the output of a pretty-printing function that has been
+   called before. For instance,
+   [
+    Pervasives.print_string "<";
+    Format.print_string "PRETTY";
+    Pervasives.print_string ">";
+    Format.print_string "TEXT";
+   ]
+   leads to output [<>PRETTYTEXT].
 
-   Warning: the material output by the following functions is delayed
-   in the pretty-printer queue in order to compute the proper line
-   splitting. Hence, you should not mix calls to the printing functions
-   of the basic I/O system with calls to the functions of this module:
-   this could result in some strange output seemingly unrelated with
-   the evaluation order of printing commands.
 *)
 
-(** {6 Boxes} *)
+(** {6 Pretty-printing boxes} *)
+
+(** The pretty-printing engine uses the concepts of pretty-printing box and
+  break hint to drive the indentation and the line splitting behavior of the
+  pretty-printer.
+
+  Each different pretty-printing box kind introduces a specific line splitting
+  policy:
+- within an {e horizontal} box, there is no line splitting,
+- within a {e vertical} box, every break hint splits the line,
+- within an {e horizontal/vertical} box there is no line splitting if the
+  box fits on the current line, otherwise every break hint splits the line,
+- within an {e compacting} box, a break hint never splits the line, unless
+  there is no more room on the current line.
+
```

Comment:
You're right, this is often unclear for many beginners. I proposed to
rephrase as follows:

Each different pretty-printing box kind introduces a specific line splitting
policy:
- within an {e horizontal} box, break hints never split the line (but the line
  may be split in a box nested deeper),
- within a {e vertical} box, break hints always split the line,
- within an {e horizontal/vertical} box, if the box fits on the current line
  then break hints never split the line, otherwise break hint always split the
  line,
- within an {e compacting} box, a break hint never splits the line,
  unless there is no more room on the current line.

Note that line splitting policy is box specific: the policy of a box does not
rule the policy of inner boxes. For instance, if a vertical box is nested in
an horizontal box, all break hints within the vertical box will split the
line.

> > -(** {6 Boxes} *)
> > +(** {6 Pretty-printing boxes} *)
> > +
> > +(** The pretty-printing engine uses the concepts of pretty-printing box and
> > +  break hint to drive the indentation and the line splitting behavior of the
> > +  pretty-printer.
> > +
> > +  Each different pretty-printing box kind introduces a specific line splitting
> > +  policy:
> > +- within an {e horizontal} box, there is no line splitting,
> > +- within a {e vertical} box, every break hint splits the line,
> > +- within an {e horizontal/vertical} box there is no line splitting if the
> > +  box fits on the current line, otherwise every break hint splits the line,
> > +- within an {e compacting} box, a break hint never splits the line, unless
> > +  there is no more room on the current line.
> > +
> 
> You should stress that the break hints are the one 'directly within' the box not the one inside inner boxes.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/ocaml/ocaml/pull/615/files/d49e17f82307d7cf396cb6cc27b9fa5ddff0cefd#r66835044